hypothesis-python/src/hypothesis/internal/conjecture/dfa/__init__.py

```diff
@@ -61,55 +84,184 @@ def matches(self, s):
             i = self.transition(i, c)
         return self.is_accepting(i)
 
+    @cached
+    def max_length(self, i):
+        """Returns the maximum length of a string that is
+        accepted when starting from i."""
+        if self.is_dead(i):
+            return 0
+        if i in self.reachable(i):
+            return inf
+        next_states = {self.max_length(j) for _, j in self.transitions(i)}
+        if next_states:
+            return 1 + max(next_states)
+        else:
+            assert self.is_accepting(i)
+            return 0
+
+    @cached
+    def count_strings(self, i, k):
+        """Returns the number of strings of length ``k``
+        that are accepted when starting from state ``i``."""
+        assert k >= 0
+        if k == 0:
+            if self.is_accepting(i):
+                return 1
+            else:
+                return 0
+        if k > self.max_length(i):
+            return 0
+        return sum(self.count_strings(j, k - 1) for _, j in self.transitions(i))
+
+    @cached
+    def reachable(self, i):
+        """Returns the set of all states reachable
+        by traversing some non-empty string starting from
+        state i."""
+        reached = set()
+        seen = set()
+
+        queue = deque([i])
+
+        while queue:
+            j = queue.popleft()
+            if j in seen:
+                continue
+            seen.add(j)
```

Comment:
Whoops, this is what I get for using “suggest change” and then not including the whole change. The point of adding `if k != i` was that we can now get rid of `seen` entirely (as shown by the coverage failure).

```suggestion
        queue = deque([i])

        while queue:
            j = queue.popleft()
```